### PR TITLE
Bugfix/issue 133 escape characters in file names

### LINKED ISSUE #133: Certain characters in the key lead to breakage
As the key is mapped to a file name in a relatively straightforward fashion, many characters can lead to non-working object access. On Windows, for example, the colon is not supported, although it is valid in a key.

There are several approaches to dealing with this problem, e.g. 
* checking whether all characters are "safe" and hashing and encoding the whole key to a hex string if they aren't
* escaping all "dangerous" characters (e.g. using URL-Encoding - the percent-character is safe on all relevant file systems)
* a more sophisticated escaping method

### PR DESCRIPTION
**Pull request by hennejg**
## Description
This PR fixes #133 as well as unit test failures under Windows

The failing tests were related to a slash-flipping-issue in `FileStore.getS3Objects` and have been fixed on the go. For the individual fix see 569bf3961592bee1f.

## Related Issue
#133 
#131 

## Tasks

- [x] I have signed the [CLA](http://adobe.github.io/cla.html).
- [x] I have written tests and verified that they fail without my change.

### CODE REVIEW

**Comment by hennejg** [comment]
Sorry, there might be another corner case lurking. Please refrain from merging this PR just yet.

**Comment by hennejg** [comment]
Yes, indeed, there were more problems: the `List Objects` and `List Objects V2` APIs have an optional `encoding-type` parameter which wasn't properly honored yet. I amended the pull requests with support for this parameter as well as accompanying test cases.

File: server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java [hunk 1]
```diff
@@ -1214,10 +1255,83 @@ private void getObjectWithRange(final HttpServletResponse response, final Range
...
+   * and macOS.
+   * 
+   * <p>The escaping is based on a modified URL encoding scheme (using 16 instead of 32 bits) 
+   * and uses different rules which characters to escape.
+   * 
+   * @param objectName the object name to encode
+   * @return encoded key
+   */
+  // @VisibleForTesting
+  static String objectNameToFileName(String objectName) {
```
Comment (agudian) [inline server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java]: The two conversion methods here are good candidates to be extracted into other classes. 

I'll merge the PR anyway, as we've neglected this one for too long, but if you are still up to it, feel free to refactor a bit in a new PR.. ;)

Comment (hennejg) [reply server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java]: Can do - that would give me the opportunity to correct the method comment for `objectNameToFileName` whhich should have contained "using 16 instead of 8 bits", of course.
What were you thinking of, a separate utility class below `com.adobe.testing.s3mock.util`?
